core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -303,6 +303,293 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
+
+    const sycl::range<1> local_range(block_size);
+    // REMEMBER swap local x<->y to be conforming with Cuda/HIP implementation
+    const sycl::range<1> global_range(nwork * block_size);
+    const sycl::nd_range<1> range{global_range, local_range};
+
+    const size_t wgroup_size = range.get_local_range().size();
+    size_t size              = range.get_global_range().size();
+    const auto init_size =
+        std::max<std::size_t>((size + wgroup_size - 1) / wgroup_size, 1);
+    const auto& selected_reducer = ReducerConditional::select(functor, reducer);
+    const unsigned int value_count =
+        FunctorValueTraits<ReducerTypeFwd, WorkTagFwd>::value_count(
+            selected_reducer);
+    const auto results_ptr = static_cast<pointer_type>(sycl::malloc_shared(
+        sizeof(value_type) * std::max(value_count, 1u) * init_size, q));
+    // FIXME_SYCL without this we are running into a race condition
+    const auto results_ptr2 = static_cast<pointer_type>(sycl::malloc_shared(
+        sizeof(value_type) * std::max(value_count, 1u) * init_size, q));
+
+    // If size<=1 we only call init(), the functor and possibly final once
+    // working with the global scratch memory but don't copy back to
+    // m_result_ptr yet.
+    if (size <= 1) {
+      q.submit([&](sycl::handler& cgh) {
+        cgh.single_task([=]() {
+          const auto& selected_reducer = ReducerConditional::select(
+              static_cast<const FunctorType&>(functor),
+              static_cast<const ReducerType&>(reducer));
+          reference_type update =
+              ValueInit::init(selected_reducer, results_ptr);
+          if (size == 1) {
+            Kokkos::Impl::Reduce::DeviceIterateTile<
+                Policy::rank, BarePolicy, Functor, typename Policy::work_tag,
+                reference_type>(policy, functor, update, {1, 1, 1}, {0, 0, 0},
+                                {0, 0, 0})
+                .exec_range();
+          }
+          if constexpr (ReduceFunctorHasFinal<FunctorType>::value)
+            FunctorFinal<FunctorType, WorkTag>::final(
+                static_cast<const FunctorType&>(functor), results_ptr);
+        });
+      });
+      m_space.fence();
+    }
+
+    // Otherwise, we perform a reduction on the values in all workgroups
+    // separately, write the workgroup results back to global memory and recurse
+    // until only one workgroup does the reduction and thus gets the final
+    // value.
+    bool first_run = true;
+    while (size > 1) {
+      auto n_wgroups = (size + wgroup_size - 1) / wgroup_size;
+      q.submit([&](sycl::handler& cgh) {
+        sycl::accessor<value_type, 1, sycl::access::mode::read_write,
+                       sycl::access::target::local>
+            local_mem(sycl::range<1>(wgroup_size) * std::max(value_count, 1u),
+                      cgh);
+
+        const BarePolicy bare_policy = m_policy;
+
+        cgh.parallel_for(range, [=](sycl::nd_item<1> item) {
```

Comment:
We may wanna do this in two different parallel for calls instead of having the if(first_run ) {} else {} inside the kernel. Effectively you are paying the full register cost for every subsequent call too.

Reply:
Yes, I was also thinking about splitting it but decided to rather try to keep it similar to the other SYCL `parallel_reduce` implementations for now.
I prefer optimizing and maybe fusing the `parallel_reduce` implementations some more once they are all merged.

Reply:
ok